File: ruby/exmo.rb
require 'net/https'
require "json"

module Exmo
  class Error < StandardError
    attr_reader :object

    def initialize(object)
      @object = object
    end
  end

  class API
    class << self
      KEY = "K-b123456"      # TODO replace with your api key from settings page
      SECRET = "S-e123456"   # TODO replace with your api secret from settings page

      def api_query(method, params = nil)
        raise ArgumentError unless method.is_a?(String) || method.is_a?(Symbol)

        params = {} if params.nil?
        params['nonce'] = nonce

        uri = URI.parse(['https://api.exmo.com/v1', method].join('/'))

        post_data = URI.encode_www_form(params)

        digest = OpenSSL::Digest.new('sha512')
        sign = OpenSSL::HMAC.hexdigest(digest, SECRET, post_data)

        headers = {
          'Sign' => sign,
          'Key'  => KEY
        }

        req = Net::HTTP::Post.new(uri.path, headers)
        req.body = post_data
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true if uri.scheme == 'https'
        response = http.request(req)

        unless response.code == '200'
          raise Exmo::Error.new(__method__), ['http error:', response.code].join(' ')
        end

        result = response.body.to_s

        unless result.is_a?(String) && valid_json?(result)
          raise Exmo::Error.new(__method__), "Invalid json"
        end

        JSON.load result
      end

      private

      def valid_json?(json)
        JSON.parse(json)
        true
      rescue
        false
      end

      def nonce
        Time.now.strftime("%s%6N")
      end
    end
  end
end



puts "%s" % Exmo::API.api_query('user_info').inspect



# puts "%s" % Exmo::API.api_query('trades', pair: 'DOGE_USD').inspect
# puts "%s" % Exmo::API.api_query('order_trades', order_id: 12345).inspect
# puts "%s" % Exmo::API.api_query('order_book', pair: 'DOGE_USD', limit: 10).inspect
# puts "%s" % Exmo::API.api_query('ticker').inspect
# puts "%s" % Exmo::API.api_query('pair_settings').inspect
# puts "%s" % Exmo::API.api_query('currency').inspect

# order_params = {
#      pair: 'DOGE_USD',
#      quantity: 100,
#      price: 99,
#      type: 'sell'
#    }
# puts "%s" % Exmo::API.api_query('order_create', order_params).inspect

# puts "%s" % Exmo::API.api_query('user_open_orders').inspect
# puts "%s" % Exmo::API.api_query('order_cancel', order_id: 12345).inspect
# puts "%s" % Exmo::API.api_query('user_cancelled_orders', limit: 100, offset: 0).inspect

# puts "%s" % Exmo::API.api_query('order_trades', order_id: 12345).inspect
# puts "%s" % Exmo::API.api_query('required_amount', pair: 'BTC_USD', quantity: 10).inspect
# puts "%s" % Exmo::API.api_query('deposit_address').inspect

File: golang/exmo.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha512"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

type ApiResponse map[string]interface{}
type ApiParams map[string]string

func api_query(method string, params ApiParams) (ApiResponse, error) {
	key := ""       // TODO replace with your api key from profile page
	secret := ""    // TODO replace with your api secret from profile page

	post_params := url.Values{}
	post_params.Add("nonce", nonce())
	if params != nil {
		for key, value := range params {
			post_params.Add(key, value)
		}
	}
	post_content := post_params.Encode()

	sign := do_sign(post_content, secret)

	req, _ := http.NewRequest("POST", "https://api.exmo.com/v1/"+method, bytes.NewBuffer([]byte(post_content)))
	req.Header.Set("Key", key)
	req.Header.Set("Sign", sign)
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Add("Content-Length", strconv.Itoa(len(post_content)))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.Status != "200 OK" {
		return nil, errors.New("http status: " + resp.Status)
	}

	body, err1 := ioutil.ReadAll(resp.Body)
	if err1 != nil {
		return nil, err1
	}

	var dat map[string]interface{}
	err2 := json.Unmarshal([]byte(body), &dat)
	if err2 != nil {
		return nil, err2
	}

	if result, ok := dat["result"]; ok && result.(bool) != true {
		return nil, errors.New(dat["error"].(string))
	}

	return dat, nil
}

func nonce() string {
	return fmt.Sprintf("%d", time.Now().UnixNano())
}

func do_sign(message string, secret string) string {
	mac := hmac.New(sha512.New, []byte(secret))
	mac.Write([]byte(message))
	return fmt.Sprintf("%x", mac.Sum(nil))
}

func main() {
	result, err := api_query("user_info", nil)
	if err != nil {
		fmt.Printf("api error: %s\n", err.Error())
	} else {
		fmt.Printf("api result: %v\n", result)
	}

	fmt.Printf("-------------\n")

	result1, err1 := api_query("user_trades", ApiParams{"pair": "BTC_USD"})
	if err1 != nil {
		fmt.Printf("api error: %s\n", err1.Error())
	} else {
		fmt.Printf("api result: %v\n", result1)
	}
}
